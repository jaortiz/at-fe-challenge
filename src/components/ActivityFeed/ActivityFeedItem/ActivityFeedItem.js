import React from "react";
import PropTypes from "prop-types";
import reactStringReplace from "react-string-replace";
import { ActivityItem, ActivityLink } from "./ActivityFeedItem.style";

/**
 * Replaces the template string references with their actual object/content references
 *
 * @param {*} template
 * @param {*} templateObj
 * @param {*} task
 * @param {*} users
 */
const formatTemplate = (
  template,
  templateObj,
  task,
  users,
  handleHoverEnter,
  handleHoverLeave
) => {
  let formattedTemplate = reactStringReplace(template, templateObj, () => {
    const id = /\d+/.exec(templateObj)[0];
    if (/profiles/.test(templateObj)) {
      return (
        <ActivityLink
          key={id}
          href={`/users/${users[id].slug}`}
          onMouseEnter={() => handleHoverEnter(`/users/${users[id].slug}`)}
          onMouseLeave={() => handleHoverLeave()}
        >
          {users[id].abbreviated_name}
        </ActivityLink>
      );
    } else if (/task/.test(templateObj)) {
      return (
        <ActivityLink
          key={id}
          href={`/tasks/${task.slug}`}
          onMouseEnter={() => handleHoverEnter(`/tasks/${task.slug}`)}
          onMouseLeave={() => handleHoverLeave()}
        >
          {task.name}
        </ActivityLink>
      );
    }
  });
  return formattedTemplate;
};

const ActivityFeedItem = ({
  activity,
  task,
  users,
  handleHoverEnter,
  handleHoverLeave
}) => {
  if (!activity || !task || !users) return null;

  let template = activity.template;
  const templateObjects = template.match(/{[^}]*\}/g);

  templateObjects.forEach(templateObj => {
    template = formatTemplate(
      template,
      templateObj,
      task,
      users,
      handleHoverEnter,
      handleHoverLeave
    );
  });

  return <ActivityItem>{template}</ActivityItem>;
};

ActivityFeedItem.propTypes = {
  activity: PropTypes.object.isRequired,
  task: PropTypes.object.isRequired,
  users: PropTypes.object.isRequired
};

export default ActivityFeedItem;
