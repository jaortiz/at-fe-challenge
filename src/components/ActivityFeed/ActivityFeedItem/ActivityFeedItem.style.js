import styled from "styled-components";

export const ActivityItem = styled.li`
  padding: 2rem;
  border-bottom: 1px solid #e8e8e8;
  text-transform: uppercase;
  color: #829094;
  font-weight: 600;
`;

export const ActivityLink = styled.a`
  color: #008fb4;
  text-decoration: none;
  text-transform: none;

  &:hover {
    border-bottom: 1px solid;
  }
`;
