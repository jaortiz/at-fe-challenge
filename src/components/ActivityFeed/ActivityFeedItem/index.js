export { default as ActivityFeedItem } from "./ActivityFeedItem";
