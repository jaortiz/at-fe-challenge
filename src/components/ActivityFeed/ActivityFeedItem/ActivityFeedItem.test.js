import React from "react";
import { shallow } from "enzyme";
import ActivityFeedItem from "./ActivityFeedItem";

describe("ActivityFeedItem", () => {
  const activity = {
    created_at: "2015-06-24T16:36:15+10:00",
    template: "{ profiles:1 } posted a task { task:1 }",
    event: "posted",
    task_id: 1,
    profile_ids: [1]
  };

  const task = {
    id: 1,
    name: "Test Task",
    slug: "test-slug"
  };

  const users = {
    "1": {
      id: 1,
      abbreviated_name: "Jeremy O",
      slug: "jeremy-o"
    }
  };

  it("renders null with no props", () => {
    const wrapper = shallow(<ActivityFeedItem />);
    expect(wrapper.getElement()).toBe(null);
  });

  it("renders correctly given props", () => {
    const mock = jest.fn();
    const wrapper = shallow(
      <ActivityFeedItem
        activity={activity}
        task={task}
        users={users}
        handleHoverEnter={mock}
        handleHoverLeave={mock}
      />
    );

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.text()).toEqual("Jeremy O posted a task Test Task");
  });
});
