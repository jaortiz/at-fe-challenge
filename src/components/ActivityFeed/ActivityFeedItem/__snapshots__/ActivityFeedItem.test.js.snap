// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ActivityFeedItem renders correctly given props 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ActivityFeedItem
    activity={
      Object {
        "created_at": "2015-06-24T16:36:15+10:00",
        "event": "posted",
        "profile_ids": Array [
          1,
        ],
        "task_id": 1,
        "template": "{ profiles:1 } posted a task { task:1 }",
      }
    }
    handleHoverEnter={[MockFunction]}
    handleHoverLeave={[MockFunction]}
    task={
      Object {
        "id": 1,
        "name": "Test Task",
        "slug": "test-slug",
      }
    }
    users={
      Object {
        "1": Object {
          "abbreviated_name": "Jeremy O",
          "id": 1,
          "slug": "jeremy-o",
        },
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        "",
        <ForwardRef
          href="/users/jeremy-o"
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          Jeremy O
        </ForwardRef>,
        " posted a task ",
        <ForwardRef
          href="/tasks/test-slug"
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          Test Task
        </ForwardRef>,
        "",
      ],
    },
    "ref": null,
    "rendered": Array [
      "",
      Object {
        "instance": null,
        "key": "1",
        "nodeType": "function",
        "props": Object {
          "children": "Jeremy O",
          "href": "/users/jeremy-o",
          "onMouseEnter": [Function],
          "onMouseLeave": [Function],
        },
        "ref": null,
        "rendered": "Jeremy O",
        "type": Object {
          "$$typeof": Symbol(react.forward_ref),
          "attrs": Array [],
          "componentStyle": ComponentStyle {
            "componentId": "sc-bwzfXH",
            "isStatic": true,
            "rules": Array [
              "
  color: #008fb4;
  text-decoration: none;
  text-transform: none;

  &:hover {
    border-bottom: 1px solid;
  }
",
            ],
          },
          "displayName": "styled.a",
          "foldedComponentIds": Array [],
          "render": [Function],
          "styledComponentId": "sc-bwzfXH",
          "target": "a",
          "toString": [Function],
          "warnTooManyClasses": [Function],
          "withComponent": [Function],
        },
      },
      " posted a task ",
      Object {
        "instance": null,
        "key": "1",
        "nodeType": "function",
        "props": Object {
          "children": "Test Task",
          "href": "/tasks/test-slug",
          "onMouseEnter": [Function],
          "onMouseLeave": [Function],
        },
        "ref": null,
        "rendered": "Test Task",
        "type": Object {
          "$$typeof": Symbol(react.forward_ref),
          "attrs": Array [],
          "componentStyle": ComponentStyle {
            "componentId": "sc-bwzfXH",
            "isStatic": true,
            "rules": Array [
              "
  color: #008fb4;
  text-decoration: none;
  text-transform: none;

  &:hover {
    border-bottom: 1px solid;
  }
",
            ],
          },
          "displayName": "styled.a",
          "foldedComponentIds": Array [],
          "render": [Function],
          "styledComponentId": "sc-bwzfXH",
          "target": "a",
          "toString": [Function],
          "warnTooManyClasses": [Function],
          "withComponent": [Function],
        },
      },
      "",
    ],
    "type": Object {
      "$$typeof": Symbol(react.forward_ref),
      "attrs": Array [],
      "componentStyle": ComponentStyle {
        "componentId": "sc-bdVaJa",
        "isStatic": true,
        "rules": Array [
          "
  padding: 2rem;
  border-bottom: 1px solid #e8e8e8;
  text-transform: uppercase;
  color: #829094;
  font-weight: 600;
",
        ],
      },
      "displayName": "styled.li",
      "foldedComponentIds": Array [],
      "render": [Function],
      "styledComponentId": "sc-bdVaJa",
      "target": "li",
      "toString": [Function],
      "warnTooManyClasses": [Function],
      "withComponent": [Function],
    },
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          "",
          <ForwardRef
            href="/users/jeremy-o"
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
          >
            Jeremy O
          </ForwardRef>,
          " posted a task ",
          <ForwardRef
            href="/tasks/test-slug"
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
          >
            Test Task
          </ForwardRef>,
          "",
        ],
      },
      "ref": null,
      "rendered": Array [
        "",
        Object {
          "instance": null,
          "key": "1",
          "nodeType": "function",
          "props": Object {
            "children": "Jeremy O",
            "href": "/users/jeremy-o",
            "onMouseEnter": [Function],
            "onMouseLeave": [Function],
          },
          "ref": null,
          "rendered": "Jeremy O",
          "type": Object {
            "$$typeof": Symbol(react.forward_ref),
            "attrs": Array [],
            "componentStyle": ComponentStyle {
              "componentId": "sc-bwzfXH",
              "isStatic": true,
              "rules": Array [
                "
  color: #008fb4;
  text-decoration: none;
  text-transform: none;

  &:hover {
    border-bottom: 1px solid;
  }
",
              ],
            },
            "displayName": "styled.a",
            "foldedComponentIds": Array [],
            "render": [Function],
            "styledComponentId": "sc-bwzfXH",
            "target": "a",
            "toString": [Function],
            "warnTooManyClasses": [Function],
            "withComponent": [Function],
          },
        },
        " posted a task ",
        Object {
          "instance": null,
          "key": "1",
          "nodeType": "function",
          "props": Object {
            "children": "Test Task",
            "href": "/tasks/test-slug",
            "onMouseEnter": [Function],
            "onMouseLeave": [Function],
          },
          "ref": null,
          "rendered": "Test Task",
          "type": Object {
            "$$typeof": Symbol(react.forward_ref),
            "attrs": Array [],
            "componentStyle": ComponentStyle {
              "componentId": "sc-bwzfXH",
              "isStatic": true,
              "rules": Array [
                "
  color: #008fb4;
  text-decoration: none;
  text-transform: none;

  &:hover {
    border-bottom: 1px solid;
  }
",
              ],
            },
            "displayName": "styled.a",
            "foldedComponentIds": Array [],
            "render": [Function],
            "styledComponentId": "sc-bwzfXH",
            "target": "a",
            "toString": [Function],
            "warnTooManyClasses": [Function],
            "withComponent": [Function],
          },
        },
        "",
      ],
      "type": Object {
        "$$typeof": Symbol(react.forward_ref),
        "attrs": Array [],
        "componentStyle": ComponentStyle {
          "componentId": "sc-bdVaJa",
          "isStatic": true,
          "rules": Array [
            "
  padding: 2rem;
  border-bottom: 1px solid #e8e8e8;
  text-transform: uppercase;
  color: #829094;
  font-weight: 600;
",
          ],
        },
        "displayName": "styled.li",
        "foldedComponentIds": Array [],
        "render": [Function],
        "styledComponentId": "sc-bdVaJa",
        "target": "li",
        "toString": [Function],
        "warnTooManyClasses": [Function],
        "withComponent": [Function],
      },
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
